File: macroslib/tests/expectations/null_annotation_java.java
r##"package org.example;
import android.support.annotation.NonNull;

public final class TrackInfo {

    private TrackInfo() {}

    public final @NonNull java.util.Date start_time() {
        long ret = do_start_time(mNativeObj);
        java.util.Date convRet = new java.util.Date(ret);

        return convRet;
    }
    private static native long do_start_time(long self);

    public final @NonNull java.util.OptionalLong end_time() {
        java.util.OptionalLong ret = do_end_time(mNativeObj);

        return ret;
    }
    private static native @NonNull java.util.OptionalLong do_end_time(long self);"##;

r#"package org.example;
import android.support.annotation.NonNull;

public final class Boo {

    public Boo() {
        mNativeObj = init();
    }
    private static native long init();

    public final @NonNull String latDirection() {
        String ret = do_latDirection(mNativeObj);

        return ret;
    }
    private static native @NonNull String do_latDirection(long self);"#;

r#"package org.example;
import android.support.annotation.NonNull;

public final class Foo {

    public Foo(double a0, double a1, double a2, double a3, double a4) {
        mNativeObj = init(a0, a1, a2, a3, a4);
    }
    private static native long init(double a0, double a1, double a2, double a3, double a4);

    public final @NonNull Boo dropPoint() {
        long ret = do_dropPoint(mNativeObj);
        Boo convRet = new Boo(InternalPointerMarker.RAW_PTR, ret);

        return convRet;
    }
    private static native long do_dropPoint(long self);"#;

r#"package org.example;
import android.support.annotation.NonNull;

public final class BooList {

    private BooList() {}

    public final @NonNull Boo [] getBooList() {
        Boo [] ret = do_getBooList(mNativeObj);

        return ret;
    }
    private static native @NonNull Boo [] do_getBooList(long self);"#;

File: macroslib/tests/expectations/null_annotation_java.rs
foreign_typemap!(
    ($p:r_type) DateTime<Utc> => jlong {
        $out = $p.timestamp_millis();
    };
    ($p:f_type, option = "NoNullAnnotations", unique_prefix = "/*chrono*/") => "/*chrono*/java.util.Date" "$out = new java.util.Date($p);";
    ($p:f_type, option = "NullAnnotations", unique_prefix = "/*chrono*/") => "/*chrono*/@NonNull java.util.Date" "$out = new java.util.Date($p);";
);

foreigner_class!(
    class TrackInfo {
        self_type TrackInfo;
        private constructor = empty;
        fn TrackInfo::start_time(&self) -> DateTime<Utc>;
        fn end_time(&self) -> Option<i64>;
    }
);

foreigner_class!(class Boo {
    self_type Boo;
    constructor Boo::default() -> Boo;
    fn latDirection(&self) -> &str;
});

foreigner_class!(class Foo {
    self_type Foo;
    constructor Foo::new(_: f64, _: f64, _: f64, _: f64, _: f64) -> Foo;
    fn Foo::drop_point(&self) -> Boo; alias dropPoint;
});

foreigner_class!(
    class BooList {
        self_type BooList;
        private constructor = empty;
        fn getBooList(&self) -> Vec<Boo>;
    }
);
